zerver/tests/test_messages.py

```diff
@@ -2081,3 +2083,226 @@ def test_delete_message_second_time(self):
+        recipient_list  = [self.example_email("hamlet"), self.example_email("iago"), self.example_email('cordelia')]
+        for email in recipient_list:
+            self.subscribe_to_stream(email, "Denmark")
+
+        sender = self.example_email('iago')
+        stream_name = 'Denmark'
+        subject = 'foo'
+
+        def send_stream_message(content):
+            # type: () -> None
+            self.send_message(sender, stream_name, Recipient.STREAM,
+                              content, subject)
+
+        def send_personal_message(content):
+            # type: () -> None
+            self.send_message(sender, self.example_email("hamlet"),
+                              Recipient.PERSONAL, content)
+
+        long_term_idle_user = self.soft_deactivate_user('hamlet')
+
+        # Test that sending a message to a stream with soft deactivated user
+        # doesn't end up creating UserMessage row for deactivated user.
+        normal_user_msg_count = len(self.get_user_ums(self.example_user('cordelia')))
+        soft_deactivated_user_msg_count = len(self.get_user_ums(self.example_user('hamlet')))
+        send_stream_message('Test Message 1')
+        self.assertNotEqual(self.get_user_ums(long_term_idle_user)[-1].message.content, u'Test Message 1')
+        self.assertEqual(len(self.get_user_ums(self.example_user('hamlet'))), soft_deactivated_user_msg_count)
+        self.assertEqual(len(self.get_user_ums(self.example_user('cordelia'))), normal_user_msg_count + 1)
+        self.assertEqual(self.get_user_ums(self.example_user('cordelia'))[-1].message.content, u'Test Message 1')
+
+        # Test sending a private message to soft deactivated user creates
+        # UserMessage row.
+        soft_deactivated_user_msg_count = len(self.get_user_ums(self.example_user('hamlet')))
+        send_personal_message('Test PM')
+        self.assertEqual(len(self.get_user_ums(self.example_user('hamlet'))), soft_deactivated_user_msg_count + 1)
+        self.assertEqual(self.get_user_ums(self.example_user('hamlet'))[-1].message.content, u'Test PM')
+
+        # Test UserMessage row is only created while user is deactivated if
+        # user itself is mentioned.
+        normal_user_msg_count = len(self.get_user_ums(self.example_user('cordelia')))
+        soft_deactivated_user_msg_count = len(self.get_user_ums(self.example_user('hamlet')))
+        send_stream_message('Test @**King Hamlet** mention')
+        self.assertEqual(self.get_user_ums(long_term_idle_user)[-1].message.content, u'Test @**King Hamlet** mention')
+        self.assertEqual(len(self.get_user_ums(self.example_user('hamlet'))), soft_deactivated_user_msg_count + 1)
+        self.assertEqual(len(self.get_user_ums(self.example_user('cordelia'))), normal_user_msg_count + 1)
+        self.assertEqual(self.get_user_ums(self.example_user('cordelia'))[-1].message.content, u'Test @**King Hamlet** mention')
+
+        # Test UserMessage row is not created while user is deactivated if
```

Comment:
For these last several chunks of code, I would fiddle around a little bit on making the test code a bit more readable/expressive. For example, I think a simple `assert_um_count(use, count)` would go a long way.

You might also want to make local vars for hamlet and cordelia.